File: enode/src/main/java/org/enodeframework/messaging/impl/MessageHandlerProxy2.java
package org.enodeframework.messaging.impl;

import org.enodeframework.common.container.ObjectContainer;
import org.enodeframework.messaging.IMessage;
import org.enodeframework.messaging.IMessageHandlerProxy2;

import java.lang.invoke.MethodHandle;
import java.lang.reflect.Method;
import java.util.concurrent.CompletableFuture;

/**
 * @author anruence@gmail.com
 */
public class MessageHandlerProxy2 implements IMessageHandlerProxy2 {
    private Class<?> handlerType;
    private Object handler;
    private MethodHandle methodHandle;
    private Method method;
    private Class<?>[] methodParameterTypes;

    @Override
    public CompletableFuture<Void> handleAsync(IMessage message1, IMessage message2) {
        Object result;
        CompletableFuture<Void> future = new CompletableFuture<>();
        try {
            if (methodParameterTypes[0].isAssignableFrom(message1.getClass())) {
                result = methodHandle.invoke(getInnerObject(), message1, message2);
            } else {
                result = methodHandle.invoke(getInnerObject(), message2, message1);
            }
            if (result instanceof CompletableFuture) {
                return (CompletableFuture<Void>) result;
            }
            future.complete(null);
        } catch (Throwable throwable) {
            future.completeExceptionally(throwable);
        }
        return future;
    }

    @Override
    public Object getInnerObject() {
        if (handler != null) {
            return handler;
        }
        handler = ObjectContainer.INSTANCE.resolve(handlerType);
        return handler;
    }

    @Override
    public void setHandlerType(Class handlerType) {
        this.handlerType = handlerType;
    }

    @Override
    public void setMethodHandle(MethodHandle methodHandle) {
        this.methodHandle = methodHandle;
    }

    @Override
    public Method getMethod() {
        return method;
    }

    @Override
    public void setMethod(Method method) {
        this.method = method;
        methodParameterTypes = method.getParameterTypes();
    }
}

File: enode/src/main/java/org/enodeframework/messaging/impl/MessageHandlerProxy2.kt
package org.enodeframework.messaging.impl

import org.enodeframework.messaging.IMessage
import org.enodeframework.messaging.IMessageHandlerProxy2
import java.lang.invoke.MethodHandle
import java.lang.reflect.Method
import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn
import kotlin.reflect.jvm.kotlinFunction

/**
 * @author anruence@gmail.com
 */
class MessageHandlerProxy2 : IMessageHandlerProxy2 {
    private lateinit var innerObject: Any
    private lateinit var methodHandle: MethodHandle
    private lateinit var method: Method

    override suspend fun handleAsync(message1: IMessage, message2: IMessage) {
        val methodParameterTypes = method.parameterTypes
        if (method.kotlinFunction?.isSuspend == true) {
            if (methodParameterTypes[0].isAssignableFrom(message1.javaClass)) {
                invokeSuspend(getInnerObject(), message1, message2)
            } else {
                invokeSuspend(getInnerObject(), message2, message1)
            }
            return
        }
        if (methodParameterTypes[0].isAssignableFrom(message1.javaClass)) {
            methodHandle.invoke(getInnerObject(), message1, message2)
        } else {
            methodHandle.invoke(getInnerObject(), message2, message1)
        }
    }

    private suspend fun invokeSuspend(obj: Any, message1: IMessage, message2: IMessage): Any? = suspendCoroutineUninterceptedOrReturn { continuation ->
        methodHandle.invoke(obj, message1, message2, continuation)
    }

    override fun getInnerObject(): Any {
        return innerObject
    }

    override fun setInnerObject(innerObject: Any) {
        this.innerObject = innerObject
    }

    override fun setMethodHandle(methodHandle: MethodHandle) {
        this.methodHandle = methodHandle
    }

    override fun getMethod(): Method {
        return method
    }

    override fun setMethod(method: Method) {
        this.method = method
    }
}